### DISCUSSION TITLE
Gracefully shutting down of NetStackImpl

### DISCUSSION THREAD

**Discussion by bdbai** [eycorsican/leaf#65] [General]
I notice that in [`NetStackImpl::new`](https://github.com/eycorsican/leaf/blob/7f3dd8ddca6d15874f3e1f592b62171cfa26a604/leaf/src/proxy/tun/netstack/stack_impl.rs#L57) there are a few `tokio::spawn` calls with a loop inside their async blocks. Some of the spawned tasks seems to run forever such as [the timer](https://github.com/eycorsican/leaf/blob/7f3dd8ddca6d15874f3e1f592b62171cfa26a604/leaf/src/proxy/tun/netstack/stack_impl.rs#L88-L96).

My suggestion is to wrap these tasks with [futures::future::abortable](https://docs.rs/futures/0.3.12/futures/future/fn.abortable.html) and manually abort them while `drop`ping a `NetStackImpl`. Do you think it is necessary to do so?

**Comment by eycorsican** [comment]
We need a way to  close all apps/connections/inbounds/outbounds gracefully, not only the netstack.

**Reply by bdbai** [reply]
Simply making the [root Future](https://github.com/eycorsican/leaf/blob/75d28684be7ad59f5efe9e35ff4ed27420c4e153/leaf-mobile/src/lib.rs#L71) *abortable* will work I guess.

**Reply by bdbai** [reply]
Aborting these root Futures makes `block_on` return. After that, we could manually call `shutdown_background` on the runtime to stop all spawned tasks.

**Reply by eycorsican** [reply]
I'm not sure how rt.shutdown_background() works, the doc says: 

> Note however, that because we do not wait for any blocking tasks to complete, this may result in a resource leak (in that any blocking tasks are still running until they return.

**Reply by bdbai** [reply]
Do we have tasks that may **block** indefinitely?
